CH-679/CH-680 added Address

NatalliaBulyha opened the pull request:
jira [Address: create tables and entity + add data] https://cozyhometeam.atlassian.net/browse/CH-680?atlOrigin=eyJpIjoiMDViMTU4MGQ4Mjc0NGVmYWEzNjE2NGUyZWQ5NGViYTEiLCJwIjoiaiJ9

Commit:
CH-679/CH-680 changed logic to build ProductDto

jira [Address: create tables and entity + add data] https://cozyhometeam.atlassian.net/browse/CH-680?atlOrigin=eyJpIjoiMDViMTU4MGQ4Mjc0NGVmYWEzNjE2NGUyZWQ5NGViYTEiLCJwIjoiaiJ9

Commit:
CH-679/CH-680 added Address

jira [Address: create tables and entity + add data] https://cozyhometeam.atlassian.net/browse/CH-680?atlOrigin=eyJpIjoiMDViMTU4MGQ4Mjc0NGVmYWEzNjE2NGUyZWQ5NGViYTEiLCJwIjoiaiJ9

Commit:
CH-679/CH-680 added Address

jira [Address: create tables and entity + add data] https://cozyhometeam.atlassian.net/browse/CH-680?atlOrigin=eyJpIjoiMDViMTU4MGQ4Mjc0NGVmYWEzNjE2NGUyZWQ5NGViYTEiLCJwIjoiaiJ9

DariaBoo commented on src/main/java/com/cozyhome/onlineshop/validation/ValidOptionalBoolean.java:
```diff
@@ -0,0 +1,27 @@
+@Retention(RUNTIME)
+public @interface ValidOptionalBoolean {
+    String message() default "Invalid boolean field. Optional field, must be not null and contains true or false.";
+
```
it a little bit confuses that field is optional but can't be null

DariaBoo commented on src/main/java/com/cozyhome/onlineshop/userservice/model/Address.java:
```diff
@@ -0,0 +1,54 @@
+    private Integer floor;
+    private boolean withLift;
+    private LocalDateTime createdAt;
+    private LocalDateTime modifiedAt;
```
we can simple add '@EqualsAndHashCode.Exclude' for needed fields to avoid them in equals and hashcode, no need to create custome equals() and hashcode() methods

NatalliaBulyha replied:
these are not a custom methods, these are overridden methods. these methods are needed so that java knows how to compare this object. Maybe I'm missing something, we need to discuss it in more detail.

NatalliaBulyha replied:
done

DariaBoo commented on src/main/java/com/cozyhome/onlineshop/userservice/controller/AddressSecuredController.java:
```diff
@@ -0,0 +1,59 @@
+@RequestMapping("${api.secure.basePath}/address")
+public class AddressSecuredController {
+    @Value("${header.name.user-id}")
+    private String userIdName;
```
maybe to rename it to userIdAttribute?

NatalliaBulyha replied:
yes, "userIdName" is unclear name

NatalliaBulyha replied:
done

DariaBoo commented on src/main/java/com/cozyhome/onlineshop/dto/user/AddressRequest.java:
```diff
@@ -0,0 +1,33 @@
+    @ValidOptionalNumber(message = "Invalid floor. Floor must be not null and contains numbers only.")
+    private String floor;
+    @ValidOptionalBoolean
+    private String withLift;
```
why we can't use here boolean withLift?

NatalliaBulyha replied:
because the string is easier to validate.  we need to discuss it in more detail.

NatalliaBulyha replied:
done

DariaBoo commented on src/main/java/com/cozyhome/onlineshop/userservice/security/service/impl/AddressServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+        if (user.getAddresses() != null) {
+            if (user.getAddresses().contains(address)) {
+                throw new DataAlreadyExistException(String.format("User with id %s already has this address: %s",
+                        userId, address));
```
do we really need to throw an exception if user already has inputed address?

DariaBoo replied:
if (!user.getAddresses().contains(address)) {
addresses = user.getAddresses();
}

NatalliaBulyha replied:
yes, we throw exception so there are no duplicates in the list.

NatalliaBulyha replied:
we need to discuss it in more detail.

DariaBoo commented on src/main/java/com/cozyhome/onlineshop/userservice/model/Address.java:
```diff
@@ -0,0 +1,54 @@
+    private Integer apartment;
+    private Integer entrance;
+    private Integer floor;
+    private boolean withLift;
```
I propose to add a field "userId" instead of List<Address> in User model. No need to save(update) user when he adds new address, just only save new Address

NatalliaBulyha replied:
we need to discuss it in more detail.

DariaBoo commented on src/main/java/com/cozyhome/onlineshop/userservice/security/service/impl/AddressServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new DataNotFoundException(String.format("User with id = %s not found.", userId)));
+
+        Address address = Address.builder()
```
propose to use:"Address address = modelMapper.map(addressRequest, Address.class);" instead of 36-55 lines.
![image](https://github.com/NatalliaBulyha/cozy-home-monolith/assets/79281909/e4eb06d2-80f1-4544-b65a-1e2837ca8dee)

NatalliaBulyha replied:
done

DariaBoo commented on src/main/java/com/cozyhome/onlineshop/userservice/security/service/impl/AddressServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+        if (!addressRequest.getWithLift().isEmpty()) {
+            address.setWithLift(Boolean.parseBoolean(addressRequest.getWithLift()));
+        }
+        if (user.getAddresses() != null) {
```
if user do not have any address will be returned null in User model or empty list?

NatalliaBulyha replied:
we return empty list

DariaBoo commented on src/main/java/com/cozyhome/onlineshop/userservice/security/service/impl/AddressServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Override
+    public void deleteAddress(AddressIdDto addressId, String userId) {
+        Address addressFromDb = addressRepository.findById(addressId.getId())
```
propose to rename 'addressFromDb' to 'addressToDelete'

NatalliaBulyha replied:
done

Commit:
CH-679/CH-680 correction on comments

jira [Address: create tables and entity + add data] https://cozyhometeam.atlassian.net/browse/CH-680?atlOrigin=eyJpIjoiMDViMTU4MGQ4Mjc0NGVmYWEzNjE2NGUyZWQ5NGViYTEiLCJwIjoiaiJ9

Commit:
CH-679/CH-680 correction on comments

jira [Address: create tables and entity + add data] https://cozyhometeam.atlassian.net/browse/CH-680?atlOrigin=eyJpIjoiMDViMTU4MGQ4Mjc0NGVmYWEzNjE2NGUyZWQ5NGViYTEiLCJwIjoiaiJ9

NatalliaBulyha committed:
Merge branch 'main' into CH-679-CH-680

Commit:
CH-675/CH-676 resolve conflicts

jira [create tables and entities + add data] https://cozyhometeam.atlassian.net/browse/CH-676?atlOrigin=eyJpIjoiZmI3NTE1NWQ0OThlNDAxOWIyZmE1NDA4MjFjYWY2OWYiLCJwIjoiaiJ9

Commit:
changed the data in accordance with figma
